agent/consul/leader_connect.go

```diff
@@ -222,6 +267,21 @@ func (s *Server) initializeRootCA(provider ca.Provider, conf *structs.CAConfigur
 	rootCA.PrivateKeyType = commonConfig.PrivateKeyType
 	rootCA.PrivateKeyBits = commonConfig.PrivateKeyBits
 
+	interPEM, err := provider.GenerateIntermediate()
+	if err != nil {
+		return fmt.Errorf("error getting intermediate cert: %v", err)
+	}
+
+	interCA, err := parseCAIntermediate(interPEM, conf.Provider, conf.ClusterID)
+	if err != nil {
+		return fmt.Errorf("error parsing intermediate cert: %v", err)
+	}
+
+	if interCA.ID != rootCA.ID {
+		//rootCA.IntermediateCerts = append(rootCA.IntermediateCerts, interPEM)
+		rootCA.SigningKeyID = interCA.SigningKeyID
```

Comment:
This doesn't seem right - why would we need to modify the root CA's representation here to have the intermediate's signing key?

Reply:
I think this is right although not sure why the line above is commented - that seems necessary too.

@kyhavlov this is because in this case the `CARoot` represents a "trust root" that has a separate Intermediate used to Sign things from the actual root key/cert. The `SigningKeyID` needs to be the ID of the key used to _sign_ leave certs which is the intermediate key here not the root key.

Reply:
Because we don't store the intermediate cert for very long, just the root cert. And in the cache logic, when we issue a leaf cert, we check for rotation by comparing the leaf cert's signing key against the _root_ cert (see [here](https://github.com/hashicorp/consul/blob/d4443374cd872ebbef9b19265418c447233eed2f/agent/cache-types/connect_ca_leaf.go#L453-L454)). But the certs are actually issued by the intermediate CA. So the end result was that we were re-issuing every leaf cert every few seconds. Nobody noticed this behavior until I added a bunch of logging to the PCA provider (although I think one user opened an issue #4475 that looks similar).